src/charm.py

```diff
@@ -68,7 +123,75 @@ def _on_config_changed(self, _):
         # TODO: placeholder method to implement logic specific to configuration change.
         pass
 
-    def _on_leader_elected(self, _) -> None:
+    def _reconfigure(self, event) -> None:
+        """Reconfigure cluster members."""
+        # Only the leader can reconfigure.
+        if not self.unit.is_leader():
+            return
+
+        # Reconfiguration can be successful only if the cluster is initialised
+        # (i.e. first unit has bootstrap the cluster).
+        if "cluster_initialised" not in self._peers.data[self.app]:
+            return
+
+        try:
+            # Compare set of Patroni cluster members and Juju hosts
+            # to avoid the unnecessary reconfiguration.
+            if self._patroni.cluster_members == self._hosts:
+                return
+
+            logger.info("Reconfiguring cluster")
+            for member in self._hosts - self._patroni.cluster_members:
+                logger.debug("Adding %s to cluster", member)
+                self.add_cluster_member(member)
+        except NotReadyError:
+            logger.info("Deferring reconfigure: another member doing sync right now")
+            event.defer()
+
+    def add_cluster_member(self, member: str):
+        """Add member to the cluster if all members are already up and running.
+
+        Raises:
+            NotReadyError if either the new member or the current members are not ready.
+        """
+        hostname = self._get_hostname_from_unit(member)
+
+        if not self._patroni.is_all_members_ready():
+            logger.info("not all members are ready")
+            raise NotReadyError("not all members are ready")
```

Comment:
Restructure this in a similar way to my suggestion for `member_started `

Reply:
I kept it this way because the unit status is being changed to blocked in the function that calls this function.